server.js

```diff
@@ -1624,29 +1622,51 @@ module.exports.initExpress = function () {
 
   app.use(require('./middlewares/redirectEffectiveAccessDenied'));
 
+  /**
+   * Attempts to extract a numeric status code from a Postgres error object.
+   * The convention we use is to use a `ERRCODE` value of `ST###`, where ###
+   * is the three-digit HTTP status code.
+   *
+   * For example, the following exception would set a 404 status code:
+   *
+   * RAISE EXCEPTION 'Entity not found' USING ERRCODE = 'ST404';
+   *
+   * @param {any} err
+   * @returns {number | null} The extracted HTTP status code
+   */
+  function maybeGetStatusCodeFromSqlError(err) {
+    const rawCode = err?.data?.sqlError?.code;
+    if (!rawCode?.startsWith('ST')) return null;
+
+    const parsedCode = Number(rawCode.toString().substring(2));
+    if (Number.isNaN(parsedCode)) return null;
+
+    return parsedCode;
+  }
+
   // This should come first so that both Sentry and our own error page can
-  // read the error ID.
+  // read the error ID and any status code.
   app.use((err, req, res, next) => {
     const _ = require('lodash');
     const chars = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'.split('');
 
     res.locals.error_id = _.times(12, () => _.sample(chars)).join('');
 
+    err.status = err.status ?? maybeGetStatusCodeFromSqlError(err) ?? 500;
+
     next(err);
   });
 
-  if (config.sentryDsn) {
-    // We need to add our own handler here to ensure that we add the appropriate
-    // information to the current Sentry scope.
-    //
-    // Note that this is typically done by our `logResponse` middleware, but
-    // that doesn't run soon enough for the Sentry error handler to pick up.
-    app.use((err, req, res, next) => {
-      enrichSentryScope(req, res);
-      next(err);
-    });
-    app.use(Sentry.Handlers.errorHandler());
-  }
+  // We need to add our own handler here to ensure that we add the appropriate
+  // information to the current Sentry scope.
+  //
+  // Note that this is typically done by our `logResponse` middleware, but
+  // that doesn't run soon enough for the Sentry error handler to pick up.
+  app.use((err, req, res, next) => {
+    enrichSentryScope(req, res);
+    next(err);
+  });
+  app.use(Sentry.Handlers.errorHandler());
```

Comment:
Similarly here, I assume it's ok to be loading this middleware unconditionally?

Reply:
Yeah, this will be a no-op if we never initialized Sentry. Originally I was thinking it'd be nice to always check `config.sentryDsn` before trying to do something with Sentry, but Sentry is already doing the "am I initialized" check internally, so it's cleaner to just unconditionally call into Sentry and allow it to no-op when appropriate.

Reply:
Cool, thanks for confirming.